mullvad-relay-selector/tests/relay_selector.rs

```diff
@@ -0,0 +1,949 @@
+            ShadowsocksEndpointData {
+                port: 443,
+                cipher: "aes-256-gcm".to_string(),
+                password: "mullvad".to_string(),
+                protocol: TransportProtocol::Tcp,
+            },
+            ShadowsocksEndpointData {
+                port: 1234,
+                cipher: "aes-256-cfb".to_string(),
+                password: "mullvad".to_string(),
+                protocol: TransportProtocol::Udp,
+            },
+            ShadowsocksEndpointData {
+                port: 1236,
+                cipher: "aes-256-gcm".to_string(),
+                password: "mullvad".to_string(),
+                protocol: TransportProtocol::Udp,
+            },
+        ],
+    },
+    wireguard: WireguardEndpointData {
+        port_ranges: vec![
+            (53, 53),
+            (443, 443),
+            (4000, 33433),
+            (33565, 51820),
+            (52000, 60000),
+        ],
+        ipv4_gateway: "10.64.0.1".parse().unwrap(),
+        ipv6_gateway: "fc00:bbbb:bbbb:bb01::1".parse().unwrap(),
+        udp2tcp_ports: vec![],
+    },
+});
+
+// Some nifty constants.
+const UDP: TransportProtocol = TransportProtocol::Udp;
+const TCP: TransportProtocol = TransportProtocol::Tcp;
+
+// Helper functions
+fn unwrap_relay(get_result: GetRelay) -> Relay {
+    match get_result {
+        GetRelay::Wireguard { inner, .. } => match inner {
+            crate::WireguardConfig::Singlehop { exit } => exit,
+            crate::WireguardConfig::Multihop { exit, .. } => exit,
+        },
+        GetRelay::OpenVpn { exit, .. } => exit,
+        GetRelay::Custom(custom) => {
+            panic!("Can not extract regular relay from custom relay: {custom}")
+        }
+    }
+}
+
+fn unwrap_endpoint(get_result: GetRelay) -> MullvadEndpoint {
+    match get_result {
+        GetRelay::Wireguard { endpoint, .. } => endpoint,
+        GetRelay::OpenVpn { endpoint, .. } => endpoint,
+        GetRelay::Custom(custom) => {
+            panic!("Can not extract Mullvad endpoint from custom relay: {custom}")
+        }
+    }
+}
+
+fn tunnel_type(relay: &Relay) -> TunnelType {
+    match relay.endpoint_data {
+        RelayEndpointData::Openvpn | RelayEndpointData::Bridge => TunnelType::OpenVpn,
+        RelayEndpointData::Wireguard(_) => TunnelType::Wireguard,
+    }
+}
+
+/// Test whether the relay selector seems to respect the order as defined by [`RETRY_ORDER`].
+#[test]
+fn test_retry_order() {
+    // In order to for the relay queries defined by `RETRY_ORDER` to always take precedence,
+    // the user settings need to be 'neutral' on the type of relay that it wants to connect to.
+    // A default `SelectorConfig` *should* have this property, but a more robust way to guarantee
+    // this would be to create a neutral relay query and supply it to the relay selector at every
+    // call to the `get_relay` function.
+    let relay_selector = RelaySelector::from_list(SelectorConfig::default(), RELAYS.clone());
```

Comment:
`RelaySelector::from_list(SelectorConfig::default(), RELAYS.clone())` seems to be a common pattern (15 times in this file). What do you think of creating a short-hand function `default_relay_selector() -> RelaySelector` or `static DEFAULT_RELAY_SELECTOR: Lazy<RelaySelector>`?

Reply:
Sure, went ahead and added the `default_relay_selector` function :blush: